cirq/circuits/circuit_gate.py

```diff
@@ -0,0 +1,190 @@
+        return self._name
+
+    @property
+    def exp_modulus(self) -> Optional[int]:
+        """The exponential modulus of this gate.
+
+        Operations which wrap this gate will have their exponent mapped to the
+        range [-exp_modulus / 2, exp_modulus / 2) if invertible, or
+        [0, exp_modulus) otherwise.
+        """
+        return self._exp_modulus
+
+    def _num_qubits_(self) -> int:
+        return protocols.num_qubits(self.circuit)
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return protocols.qid_shape(self.circuit)
+
+    def on(self, *qubits: 'cirq.Qid') -> 'CircuitOperation':
+        return CircuitOperation(self, qubits)
+
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]
+                                      ) -> 'CircuitGate':
+        new_gate = CircuitGate()
+        new_gate._circuit = protocols.with_measurement_key_mapping(
+            self.circuit, key_map)
+        new_gate._name = self.name
+        new_gate._exp_modulus = self.exp_modulus
+        return new_gate
+
+    def __eq__(self, other) -> bool:
+        if type(other) != type(self):
+            return False
+        return (self.circuit == other.circuit and self.name == other.name and
+                self.exp_modulus == other.exp_modulus)
+
+    def __hash__(self) -> int:
+        return hash((self.circuit, self.name, self.exp_modulus))
+
+    def __pow__(self, power) -> 'CircuitGate':
+        try:
+            return CircuitGate(self.circuit**power, device=self.circuit.device)
+        except:
+            return NotImplemented
+
+    def __repr__(self) -> str:
+        base_repr = repr(self.circuit)
+        if self.name is not None:
+            base_repr += f""", name='{self.name}'"""
+        if self.exp_modulus is not None:
+            base_repr += f', exp_modulus={self.exp_modulus}'
+        return f'cirq.CircuitGate({base_repr})'
+
+    def __str__(self) -> str:
+        header = 'CircuitGate:' if self.name is None else f'{self.name}:'
+        msg_lines = str(self.circuit).split('\n')
+        msg_width = max([len(header) - 4] + [len(line) for line in msg_lines])
+        full_msg = '\n'.join([
+            '[ {line:<{width}} ]'.format(line=line, width=msg_width)
+            for line in msg_lines
+        ])
+        return header + '\n' + full_msg
+
+    def _commutes_(self, other: Any,
+                   atol: float) -> Union[None, NotImplementedType, bool]:
+        # Until Circuit supports _commutes_, this will return NotImplemented.
+        return protocols.commutes(self.circuit, other, atol=atol)
+
+    def _has_unitary_(self) -> bool:
+        return protocols.has_unitary(self.circuit)
+
+    def _decompose_(self, qubits=None) -> 'cirq.OP_TREE':
+        applied_circuit = self.circuit.unfreeze()
+        if qubits is not None and qubits != self.ordered_qubits():
+            qmap = {old: new for old, new in zip(self.ordered_qubits(), qubits)}
```

Comment:
The implicit reliance on a particular ordering here seems like it could be error prone.